docs/rfcs/027-pageserver-migration.md

```diff
@@ -0,0 +1,463 @@
+# Seamless tenant migration
+
+- Author: john@neon.tech
+- Created on 2023-08-11
+- Implemented on ..
+
+## Summary
+
+The preceding [generation numbers RFC](025-generation-numbers.md) may be thought of as "making tenant
+migration safe". Following that,
+this RFC is about how those migrations are to be done:
+
+1. Seamlessly (without interruption to client availability)
+2. Quickly (enabling faster operations)
+3. Efficiently (minimizing I/O and $ cost)
+
+These points are in priority order: if we have to sacrifice
+efficiency to make a migration seamless for clients, we will
+do so, etc.
+
+This is accomplished by introducing two high level changes:
+
+- A dual-attached state for tenants, used in a control-plane-orchestrated
+  migration procedure that preserves availability during a migration.
+- Warm secondary locations for tenants, where on-disk content is primed
+  for a fast migration of the tenant from its current attachment to this
+  secondary location.
+
+## Motivation
+
+Migrating tenants between pageservers is essential to operating a service
+at scale, in several contexts:
+
+1. Responding to a pageserver node failure by migrating tenants to other pageservers
+2. Balancing load and capacity across pageservers, for example when a user expands their
```

Comment:
How do you imagine this to help with balancing load?

I spent some time thinking about it and ended up with:

1. We still care about disk utilization, so both primary AND secondary locations should be scheduled in a way that they utilize some pretty high % of the pageserver resources. Otherwise, we may end up with under-utilised pageservers and higher costs
2. Yet, this means that migrating from primary to secondary only serves HA purpose and doesn't help with expanding / freeing disk space, i.e. if tenant has an accelerated usage growth

There are several cases I can imagine, though.

1. We setup new pageserver
2. Move secondary lcoation to it
3. Do switchover of primary location to new secondary with a lot of disk space

OR

Migrate other tenants' secondary locations nearby target tenant primary one

OR

1. Migrate other tenants' secondary locations nearby target tenant secondary one
2. Do switchover of secondary to primary

Did you think about something else?

Reply:
Yes, it's all possible variations of the above.  The general point here is that seamless migration is a necessary dependency for doing anything like this -- the migration procedure itself doesn't deliver the balancing, but it provides the building block for implementing a balancing system in the future.

In the near term, I think the most common scenario will be responding tactically to disk pressure, like this:
- Tenant X has locations on pageservers A & B
- Pageserver A is becoming dangerously full
- Transfer tenant X to use pageserver B as its primary
- Create a new secondary location on pageserver C, and wait for it to warm up
- Detach the original location on pageserver A.

Longer term, I'd imagine a more continuous reconciliation loop that looks for opportunities to improve placement (based on load as well as capacity), and also handles graceful draining of nodes to decommission.